sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/QueryPlanningTracker.scala

```diff
@@ -89,8 +90,32 @@ object QueryPlanningTracker {
   }
 }
 
+/**
+ * Callbacks after planning phase completion.
+ */
+trait QueryPlanningTrackerCallback {
```

Comment:
Actually `catalyst` is an internal package so API shouldn't live here. Also, if you're exposing this, it should have things like `@since 3.5.0` too. In addition, I think it has to be developer API that has `@DeveloperApi` annotation.

Reply:
And, in fact traits are discouraged as API because of Java compat, see https://github.com/databricks/scala-style-guide#traits-and-abstract-classes. But I am fine as it's a developer API.

Reply:
I think `private[spark]` could work for these added sql APIs, as they are inteded to be used internally from Spark Connect, maybe later also from Thriftserver.
And the whole QueryPlanningTracker would better fit in `sql-core` rather than `catalyst` (as it e.g. measures also physical planning, which is not in catalyst; and it would actually be convenient to pass QueryExecution to these callbacks, but we can't in catalyst). But moving it would break the existing uses of QueryPlanningTracker...

Reply:
Let's make these private so that we can use them from Spark Connect, but not expose this as a public API because it's not meant for that.

Reply:
- Added annotations: since, DeveloperApi
- Updated from trait to abstract class